Add DurableTask.ApplicationInsights package and DurableTelemetryModule

bachuv opened the pull request:
This PR adds a new package `DurableTask.ApplicationInsights` which has a `DurableTelemetryModule`. This telemetry module listens to the `Activity`s that are emitted and converts them to Application Insights readable information (e.g. dependency telemetry). This is necessary since the Functions Host supports Application Insights and not OpenTelemetry yet.

Related Durable Functions PR - https://github.com/Azure/azure-functions-durable-extension/pull/2411

<img width="803" alt="image" src="https://github.com/Azure/durabletask/assets/15795068/689a8751-a54b-4086-893b-5f527dfd30b0">

davidmrdavid commented on src/DurableTask.ApplicationInsights/DurableTelemetryModule.cs:
```diff
@@ -0,0 +1,104 @@
+                        dependencyHolder.Dispose();
+                    }
+                },
+                Sample = (ref ActivityCreationOptions<ActivityContext> _) => ActivitySamplingResult.AllData,
+                SampleUsingParentId = (ref ActivityCreationOptions<string> _) => ActivitySamplingResult.AllData
```
would appreciate some comments here on end-effect this has on sampling :) . I'm not immediately sure what this is achieves.

bachuv replied:
Of course! I'll add a comment. These lines are just saying that the `ActivityListener` is going to collect all of the data for any `Activity` that's sent to the "DurableTask" source.

davidmrdavid commented on src/DurableTask.ApplicationInsights/DurableTelemetryModule.cs:
```diff
@@ -0,0 +1,104 @@
+                    cts.Dispose();
+                }
+            }
+
+            GC.SuppressFinalize(this);
```
Interesting - did we find this to be necessary to avoid some perf hit or are we just being perf conscious from the start by adding this GC call?

jviau replied:
I am not certain the SuppressFinalize is needed? We have no finalizer.

davidmrdavid reviewed (commented):
Left some minor, non-urgent, comments from taking a quick look!

jviau commented on src/DurableTask.ApplicationInsights/DiagnosticActivityExtensions.cs:
```diff
@@ -0,0 +1,98 @@
+
+            return tempStatusCode.HasValue;
+        }
+        private static ActivityStatusCode? GetStatusCodeForTagValue(string statusCodeTagValue)
```
nit: please add newline

jviau commented on src/DurableTask.ApplicationInsights/DiagnosticActivityExtensions.cs:
```diff
@@ -0,0 +1,98 @@
+        public static ActivityStatusCode GetStatus(this Activity activity)
+            => s_getStatus(activity);
+
+        public static bool TryGetStatus(this Activity activity, out ActivityStatusCode statusCode, out string statusDescription)
```
I think you will also need to use reflection for the System.Diagnostics.DiagnosticSources >= 6.0.0 scenario - checking if `Activity` has `.Status` and `.StatusDescription` properties on them and using those.

jviau commented on src/DurableTask.ApplicationInsights/DurableTask.ApplicationInsights.csproj:
```diff
@@ -0,0 +1,37 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+	<Import Project="$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory),DurableTask.sln))\tools\DurableTask.props" />
```
nit: align indentation with `<PropertyGroup>` below.

jviau commented on src/DurableTask.ApplicationInsights/DurableTelemetryModule.cs:
```diff
@@ -0,0 +1,104 @@
+                ShouldListenTo = source => source.Name.StartsWith("DurableTask"),
+                ActivityStarted = activity =>
+                {
+                    var dependency = _telemetryClient.StartOperation<DependencyTelemetry>(activity);
```
Our telemetry type will depend on `activity.Kind`.

Consumer or Server -> RequestTrelemetry
Client or Producer -> DependencyTelemetry. I am not sure what the telemetry type should be here. Should we make it our own string? Like "DurableTask"?
Internal -> DependencyTelemetry, "InProc" type.

jviau commented on src/DurableTask.ApplicationInsights/DurableTelemetryModule.cs:
```diff
@@ -0,0 +1,104 @@
+                ActivityStopped = activity =>
+                {
+                    // Check for Exceptions events
+                    foreach (ActivityEvent activityEvent in activity.Events)
```
I don't think we want to keep this behavior - exceptions should be logged as `ExceptionTelemetry` independently of the span being logged.

This is because they are two different telemetry concepts: exception is an app-insights flavoring of the _log_ telemetry, and `Activity` here is a _trace_ telemetry. They have slightly different semantics around them, and we should not intertwine them like this. Particularly, traces may be sampled out - we wouldn't want exceptions logging to be ignored due to sampling.

jviau reviewed (changes requested):
@bachuv this would need to be a separate sample from `OpenTelemetrySample.csproj`, as this is using the _Application Insights_ SDK - which is different than the OTel SDK. These are independent from one another, and a customer would _typically_ have one or another and not both.

jviau commented:
Can we split out the fixes to our spans to a separate PR? The changes to client/server kind and parent/child relationships.

bachuv commented:
> Can we split out the fixes to our spans to a separate PR? The changes to client/server kind and parent/child relationships.

Yes, I moved those changes to #882

cgillum commented on samples/DistributedTraceSample/ApplicationInsights/Processor.cs:
```diff
@@ -0,0 +1,32 @@
+
+        public void Process(ITelemetry item)
+        {
+            if (item is DependencyTelemetry d && d.Target.StartsWith("127.0.0.1"))
```
Can you add comments explaining what the code in this processor is doing? I don't immediately understand why you have these `if` statements.

bachuv replied:
Yes, this is more like a template to filter out/modify the telemetry. I actually see value in checking if `d.Target.Contains("core.windows.net")` to remove the storage access spans (example below). I updated the code and added a comment.

<img width="602" alt="image" src="https://user-images.githubusercontent.com/15795068/231039390-fee94432-fa46-432a-a22c-c51656d7362b.png">

cgillum replied:
Oh sorry, you already did this. I was looking at the wrong commit/update.

cgillum commented on src/DurableTask.ApplicationInsights/DiagnosticActivityExtensions.cs:
```diff
@@ -0,0 +1,140 @@
+
+        static DiagnosticActivityExtensions()
+        {
+            s_getStatus = CreateGetStatus();
```
nit: consider inlining this in line 22.

cgillum commented on src/DurableTask.ApplicationInsights/DiagnosticActivityExtensions.cs:
```diff
@@ -0,0 +1,140 @@
+                        statusDescription = tag.Value;
+                        break;
+                    default:
+                        continue;
```
This `default: continue` is redundant and can be removed.

cgillum commented on src/DurableTask.ApplicationInsights/DiagnosticActivityExtensions.cs:
```diff
@@ -0,0 +1,140 @@
+        private static bool TryGetStatusCodeForTagValue(string statusCodeTagValue, out ActivityStatusCode statusCode)
+        {
+            ActivityStatusCode? tempStatusCode = GetStatusCodeForTagValue(statusCodeTagValue);
+            statusCode = tempStatusCode ?? default;
```
Is this `?? default` redundant?

bachuv replied:
No, I don't think so. `default` will resolve to `Unset` (first value in `ActivityStatusCode` enum).

cgillum replied:
I see. I misunderstood how the compiler was going to treat `default` in this particular expression.

cgillum commented on src/DurableTask.ApplicationInsights/DurableTelemetryModule.cs:
```diff
@@ -0,0 +1,136 @@
+
+namespace DurableTask.ApplicationInsights
+{
+    /// Telemetry Module to convert Activity events to App Insights API
```
This needs to be a proper XML Doc command with `<summary>` tags.

cgillum commented on src/DurableTask.ApplicationInsights/DurableTelemetryModule.cs:
```diff
@@ -0,0 +1,136 @@
+    /// Telemetry Module to convert Activity events to App Insights API
+    public sealed class DurableTelemetryModule : ITelemetryModule, IAsyncDisposable
+    {
+        private TelemetryClient _telemetryClient;
```
nit: this file doesn't follow the coding conventions for the rest of our durabletask code. For example, we don't prefix instance variables with `_`, we use `this.` (generally), and the `using` statements should be inside the `namespace` block. I suggest we keep this code consistent with existing style.

cgillum commented on src/DurableTask.Core/Tracing/TraceHelper.cs:
```diff
@@ -248,6 +249,7 @@ public class TraceHelper
                 string statusDescription = "";
                 if (errorPropagationMode == ErrorPropagationMode.SerializeExceptions)
                 {
+                    // Null ref here if IncludeDetails is not enabled. Maybe failedEvent.Reason is good enough?
```
I don't really understand what this comment is referring to. `IncludeDetails` is a property on `TaskOrchestrationDispatcher` that controls whether `TaskScheduledEvent`'s `Input` property gets assigned, IIRC. But the line of code below only deals with `TaskFailedEvent`, which shouldn't be impacted. Also, not sure why `FailedEvent.Reason` is being mentioned when we're not actually referencing it.

Also, we should code defensively. If we think that something can fail and result in an exception, we should handle it and do something reasonable.

jviau replied:
This was from before the refactor to use `failedEvent.Reason`, when this code was trying to extract extra details from the exception. It was null-ref'ing in my testing, which led me to add the comment to investigate it later. The comment was more a TODO for my own reminder, doesn't need to be included in this PR - especially since I believe we are using `failedEvent.Reason` now.

bachuv commented on src/DurableTask.ApplicationInsights/DiagnosticActivityExtensions.cs:
```diff
@@ -0,0 +1,140 @@
+        private static Func<Activity, StatusTuple> CreateGetStatus()
+        {
+            PropertyInfo getStatus = typeof(Activity).GetProperty("Status");
+            if (getStatus is null)
```
The default value for `Activity.Status` is `Unset` so I think we can actually remove the null check. @jviau, do you agree that it can be removed?

cgillum replied:
My question was actually about when `typeof(Activity).GetProperty("Status")` would be null. Can we add a comment explaining this?

jviau replied:
`Activity.Status` and `Activity.StatusDescription` was only added in `System.Diagnostics.DiagnosticSource/6.0.0`. We only target `5.0.0`. This code is designed to test if `6.0.0` (by virtue of the properties existing) was loaded and use the properties if they are available (via reflection - well, we compile a delegate which has direct property access performance levels). Otherwise, it will fallback to using tags to store these values.

bachuv commented on src/DurableTask.ApplicationInsights/DurableTelemetryModule.cs:
```diff
@@ -0,0 +1,136 @@
+    public sealed class DurableTelemetryModule : ITelemetryModule, IAsyncDisposable
+    {
+        private TelemetryClient _telemetryClient;
+        private ActivityListener _listener;
```
No, they're assigned in `Initialize(TelemetryConfiguration configuration)`

cgillum commented on samples/DistributedTraceSample/ApplicationInsights/Processor.cs:
```diff
@@ -0,0 +1,32 @@
+                return;
+            }
+
+            if (item is RequestTelemetry request && request.Name.StartsWith("orchestration"))
```
Why do we have this `if` condition? If it's here for a reason, can we add code comments explaining why?

cgillum commented on samples/DistributedTraceSample/ApplicationInsights/Processor.cs:
```diff
@@ -0,0 +1,32 @@
+using Microsoft.ApplicationInsights;
```
Need a copyright header on this file to be consistent with the other files in this project.

jviau commented on .gitignore:
```diff
@@ -13,6 +13,7 @@ obj/
 pkg/
 packages/
 TestResults/
+Debug/
```
This is fine to add, but I am curious what is outputting to this folder?

bachuv replied:
There were some binaries and other files that were outputted. [This commit](https://github.com/Azure/durabletask/pull/871/commits/74f3048a1d75b8aab072fb2eea26d71a63147ca0) shows the files that were deleted.

jviau commented on samples/DistributedTraceSample/ApplicationInsights/ApplicationInsightsSample.csproj:
```diff
@@ -0,0 +1,24 @@
+
+  <ItemGroup>
+    <PackageReference Include="System.Diagnostics.DiagnosticSource" Version="7.0.2" />
+    <PackageReference Include="Vio.DurableTask.Hosting" Version="2.2.1" />
```
I wrote this sample with my hosting package, mostly because I wanted to show using the module with DI and this was the quickest way to get DTFx with DI. 

But do we want to keep that this way? Do we want to start including samples that rely on external extension packages?

cgillum replied:
I don't feel strongly, but my preference would be to avoid relying on external packages unless we think they are an important part of the sample. That way users can more easily adapt samples to their own projects.

jviau commented on samples/DistributedTraceSample/ApplicationInsights/Processor.cs:
```diff
@@ -0,0 +1,40 @@
+        public void Process(ITelemetry item)
+        {
+            // Filters out storage access spans so Gantt chart is easier to read
+            if (item is DependencyTelemetry d && d.Target.Contains("core.windows.net"))
```
Consider adding a filter for azurite / emulator, which is 127.0.0.1.

But also add a comment that it is not recommended to use this processor directly, and is just for demonstrative purposes. This processor will filter out **a lot** more than what the average customer would like

jviau commented on samples/DistributedTraceSample/ApplicationInsights/appsettings.json:
```diff
@@ -0,0 +1,8 @@
+{
+  "TaskHub": {
+    "Name": "appinsightssample",
+    "ErrorPropagationMode": "UseFailureDetails",
```
note for testing / bug bash: we need to test with different error propagation modes and include details values.

jviau commented on src/DurableTask.ApplicationInsights/DiagnosticActivityExtensions.cs:
```diff
@@ -0,0 +1,147 @@
+
+        private static ActivityStatusCode GetOtelStatus(this Activity activity, out string statusDescription)
+        {
+            Debug.Assert(activity != null, "Activity should not be null");
```
`Debug.Assert` only applies during `Debug` mode (and not in `Release`). As such, you want to still handle the release error case. There are two paths for this:

1. Change this to a null check + `ArgumentNullException` if null.
2. Return `ActivityStatusCode.Unset` and `null` statusDescription when `activity == null` (_after_ the `Debug.Assert`)
    - This way we are warned up front if something is awry in debug / tests, but we won't impact something using this in production.

bachuv replied:
Thanks! I implemented the second option.

jviau commented on src/DurableTask.ApplicationInsights/DurableTelemetryModule.cs:
```diff
@@ -0,0 +1,138 @@
+            {
+                case ActivityKind.Consumer or ActivityKind.Server:
+                    RequestTelemetry request = CreateTelemetryCore<RequestTelemetry>(activity);
+                    request.Success = status != ActivityStatusCode.Error;
```
For an added layer of resiliency, can you add a check for `request.StatusCode ??= (request.Success ? "success" : "fail");`. This is to guard against `durabletask.task.status` not being preset and we never set `StatusCode` (which if not set app insights silently fails to send this up).

Also instead of the raw strings `success` and `fail` - set it to the equivalent of `durabletask.task.status` in success/fail cases.

bachuv replied:
I ended up adding the following check. I replaced `??=` with a null check because `request.ResponseCode` was never `null` and evaluated to an empty string - `""`. The `bool` cast is there for `request.Success` because it's nullable.

```
if (string.IsNullOrEmpty(request.ResponseCode))
{
    request.ResponseCode = (bool)request.Success ? "200" : "500";
}
```

jviau commented on src/DurableTask.ApplicationInsights/DurableTelemetryModule.cs:
```diff
@@ -0,0 +1,138 @@
+        async ValueTask IAsyncDisposable.DisposeAsync()
+        {
+            this.listener?.Dispose();
+            if (this.telemetryClient != null)
```
nit: this will function fine, but if we want to cross our t's and dot our i's we want to do one of the following for a proper disposal pattern:

1. Either make the `telemetryClient` and `listener` fields `readonly` and remove the two null-checks
2. Keep them non-readonly, and `null` them out during dispose, then you need to go update all other methods with a `VerifyNotDisposed` pattern to ensure they don't null-ref.
   - you also need to either perform locking or lock in the fields at the start of each method.
   - as I talk about this, it is much more complex than option 1 for not much of a benefit. You only need to do this pattern when you **must** release the references to let the GC clean them up. I'd just go with option 1.

jviau replied:
Ah right, forgot about that. Option 1 is out :).

This is fine for now, we can worry about improving dispose later.

jviau commented on src/DurableTask.Core/Tracing/TraceHelper.cs:
```diff
@@ -249,20 +249,7 @@ public class TraceHelper
-                    }
-                }
-
+                string statusDescription = failedEvent.Reason ?? "Unknown task failure";
```
Not sure I like `Unknown task failure` much. Worried customers will think it is the frameworks fault for not knowing. Can we have a generic message like "TaskActivity failed"?

Open to other suggestions as well.

jviau commented on src/DurableTask.Core/Tracing/TraceHelper.cs:
```diff
@@ -349,20 +337,7 @@ public class TraceHelper
-                    }
-                }
-
+                string statusDescription = failedEvent.Reason ?? "Unkown sub-orchestration failure";
```
Same here, can we have `Sub-orchestration failed`? Open to other suggestions.

cgillum replied:
How about "Unspecified sub-orchestration failure"? The main thing I'm hoping to communicate is that we don't have the details available to us.

jviau commented on src/DurableTask.ApplicationInsights/DurableTask.ApplicationInsights.csproj:
```diff
@@ -0,0 +1,35 @@
+
+    <ItemGroup>
+        <PackageReference Include="Microsoft.ApplicationInsights" Version="2.21.0" />
+        <PackageReference Include="Microsoft.Bcl.AsyncInterfaces" Version="5.0.0" />
```
Is this package needed?

bachuv replied:
Yes, it's needed for `IAsyncDisposable`.

jviau commented on src/DurableTask.ApplicationInsights/DurableTelemetryModule.cs:
```diff
@@ -0,0 +1,138 @@
+            foreach (KeyValuePair<string, object> item in activity.TagObjects)
+            {
+                telemetry.Properties[item.Key] = item.Value.ToString();
+                if (item.Key == "durabletask.task.status" && telemetry is RequestTelemetry request)
```
Thinking about this a bit more, can you remove this if-statement entirely. See my other comment for the alternative.

cgillum commented on samples/DistributedTraceSample/ApplicationInsights/Program.cs:
```diff
@@ -0,0 +1,126 @@
+                {
+                    services.TryAddEnumerable(ServiceDescriptor.Singleton<ITelemetryModule, DurableTelemetryModule>());
+                    services.AddSingleton(GetOrchestrationService());
+                    AddApplicationInsights(services);
```
I feel like there's a lot of setup code that we're asking developers to add and I would feel better if we could abstract it all from them. This might be related to Jacob's point about whether we need to use his library, because it seems like this sample is taking a dependency on it to do this setup. Are there alternate ways to configure this that we could consider?

jviau replied:
We aren't the ones asking developers to add this - it is how app insights SDK works. The only one we are specifically asking is the line:

``` CSharp
services.TryAddEnumerable(ServiceDescriptor.Singleton<ITelemetryModule, DurableTelemetryModule>());
```

Everything else is setup code someone using app insights would have already done. If they are using DI, they would use app insights DI extensions like we do here. If they are not using DI they would be doing whatever other way there is to construct and use app insights.

But yeah, if we do want to remove the DI library here we can - it won't make it any _less_ code (probably more), but just different setup code.

bachuv replied:
> But yeah, if we do want to remove the DI library here we can - it won't make it any less code (probably more), but just different setup code.

@jviau, do you have any pointers you can share about this approach?

jviau replied:
This page has an example of adding AppInsights via a manual service collection; https://learn.microsoft.com/en-us/azure/azure-monitor/app/worker-service#net-corenet-framework-console-application

bachuv replied:
Thanks! I'll take a look.

cgillum reviewed (commented):
A few more comments from me (I also replied to some of the previous comment threads).

cgillum reviewed (approved):
Just a couple small things.
